api/README.md

```diff
@@ -0,0 +1,264 @@
+
+func (s *Server) ComputeDouble(ctx context.Context, request *internal.ComputeDoubleRequest, version apiversion.Version) (*internal.ComputeDoubleResponse, error) {
+	in := request.Input64
+	out := 2 * in
+	
+	response := &internal.ComputeDoubleResponse{}
+
+	if sign(in) != sign(out) {
+		// overflow
+		response.Overflow = true
+	} else {
+		response.Response = out
+	}
+
+	return response, nil
+}
+
+func sign(x int64) int {
+	switch {
+	case x > 0:
+		return 1
+	case x < 0:
+		return -1
+	default:
+		return 0
+	}
+}
+```
+All the boilerplate code to:
+ * add a named pipe to the server for each version of the API group, listening for its version's requests, and replying with its version's responses
+ * convert versioned requests to internal representations
+ * and then internal responses back to versioned responses
+ * create clients to talk to that API group & versioned
+is then generated automatically using [gengo](https://github.com/kubernetes/gengo).
+
+The only caveat is that when conversions cannot be made trivially (e.g. when fields from internal and versioned `struct`s have different types), API devs need to define conversion functions. They can do that by creating an (otherwise optional) `internal/server/<api_group_name>/internal/<version>/conversion.go` file, containing functions of the form `func convert_pb_<Type>_To_internal_<Type>(in *pb.<Type>, out *internal.<Type>) error` or `func convert_internal_<Type>_To_pb_<Type>(in *internal.<Type>, out *pb.<Type>) error`; for example, in our `dummy` example above, we need to define a conversion function to account for the different fields in requests and responses from `v1alpha1` to `v1`; so `internal/server/dummy/internal/v1alpha1/conversion.go` could look like:
+```go
+func convert_pb_ComputeDoubleRequest_To_internal_ComputeDoubleRequest(in *pb.ComputeDoubleRequest, out *internal.ComputeDoubleRequest) error {
+	out.Input64 = int64(in.Input32)
+	return nil
+}
+
+func convert_internal_ComputeDoubleResponse_To_pb_ComputeDoubleResponse(in *internal.ComputeDoubleResponse, out *pb.ComputeDoubleResponse) error {
+	i := in.Response
+	if i > math.MaxInt32 || i < math.MinInt64 {
+		// overflow
+		out.Overflow = true
+	} else {
+     		out.Response32 = int32(i)
+    }
+	return nil
+}
+```
+
+## How to change the API
+
+### How to add a new API group
+
+Simply create a new `api/<api_group_name>/<version>/api.proto` file, defining your new service; then generate the Go protobuf code, and run the CSI-proxy generator to generate all the boilerplate code.
+
+FIXME: add more details on which commands to run, and which files to edit when done generating.
+
+### How to make changes to an existing API group & version
+
+Changes to existing APIs are only acceptable as long as they are backward compatible; where compatibility is understood here much [as k8s API guidelines define it](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility):
```

Comment:
I know K8s APIs allow things like adding new fields without bumping the version, but that ends up adding one more version dependency, ie "I support 1.15 version of v1 API".  I had a conversation with liggitt a while ago about CRD versioning, and he recommended that we always bump the version when we make any change to the API to remove any ambiguity.

Reply:
That is indeed a fair point, thank you for bringing it up! Changing.

Reply:
Thinking a bit more about this, that would also mean bumping versions at an alarming rate. How about we simply choose to make stable versions (ie alpha and beta) immutable once they're promoted, but alpha and beta versions simply have to stay backward compatible with themselves (ie adding fields to a non-stable version doesn't require bumping to a new version altogether)?

Otherwise, every time there is any change to an API, that would require creating a new version for it, and keeping it around for quite a while (as per the deprecation policy), and we'd fast end up with a _lot_ of versions. And although the overhead to having each version is rather low, it still adds up: one named pipe per version, plus the need to add conversion functions for all lower versions when making a breaking change that requires a conversion.

If you agree with that, I will simply change the beginning of this section to read:
```
Stable API versions (i.e. not alpha nor beta) are considered immutable, and cannot be changed.

Changes to existing alpha or beta API versions are only acceptable [...]
```

Reply:
I would expect most of the changes to happen during alpha, and much fewer once beta or GA.  And for alpha it doesn't make a difference either way since it can change release to release.  I think it's much more simpler/obvious from a client's perspective to be explicit about the api version and proxy version compatibility ("If I use v1alpha3, I know it's supported by proxy version X, Y", instead of "If I use this function/field from v1alpha1, I need to use it with proxy version X")

Reply:
> I think it's much more simpler/obvious from a client's perspective to be explicit about the api version and proxy version compatibility ("If I use v1alpha3, I know it's supported by proxy version X, Y", instead of "If I use this function/field from v1alpha1, I need to use it with proxy version X")

Absolutely, totally agree. My main concern here is simply to ensure that we don't end up maintaining an unreasonable number of API versions at any one time.

> And for alpha it doesn't make a difference either way since it can change release to release.

So are you proposing that alpha APIs offer no guarantees whatsoever, not even backward compatibility with themselves, but beta and GA APIs are immutable? If so, we could even get rid of beta versions altogether, since they don't really have much purpose?

Reply:
I'm thinking it's simpler to say they are all immutable, but whether or not the proxy needs to carry support for it depends on the designation:
- alpha: 0 releases
- beta: 9 months or 3 releases
- ga: 12 months or 3 releases

What this implies is that alpha is where most of the iteration and testing should go, and we should only promote to beta once we have confidence in the api. Even once an api is promoted to beta or GA, you can add more features with a new alpha version and have that go through beta, GA cycles again.

Reply:
Agree with the above suggestion. During alpha, as the API churns, the API will go through vXalpha1 -> vXalpha2 -> ... -> vXalphaN. Once vXbeta1 gets cut, churn will be much less and we should have at most vXbeta1/vXbeta2 and finally the stable vX.

All the Alpha named pipes + API versions can thus be dropped frequently at every release and prevent an increasing build-up. The limited number of beta pipes + API versions will be carried for a limited amount of time and so on.

Once a new Alpha for vX+1 is started, unpromoted APIs will need to be carried over manually by maintainers from vX if they were not ready to be promoted to vX Beta.